### ISSUE TITLE
Errors when loading the properties file

### ISSUE THREAD

**Issue by FrancoisPaquette** [medusa-project/cantaloupe#100]
Hi,

I'm testing the version 3.3 and I'm seeing a few errors when the properties file is being reloaded. I tried with the sample file to be sure mine wasn't problematic somehow.

1.  The FilesystemWatcher function detected a change but is generating an error. 
2. Right after, you can see the timestamp is not right, it's going from 14:38 to 14:16
3. At 14:16:34,058, you can see that two ressources are missing. But looking at their names, it's tests you probably did.

Reloading config file: /opt/Cantaloupe-3.3/cantaloupe.properties
14:38:21.591 [pool-2-thread-1] DEBUG o.a.c.c.PropertiesConfiguration - FileName set to cantaloupe.properties
14:38:21.592 [pool-2-thread-1] DEBUG o.a.c.c.PropertiesConfiguration - Base path set to /opt/Cantaloupe-3.3
14:38:21.592 [pool-2-thread-1] DEBUG o.a.c.c.ConfigurationUtils - ConfigurationUtils.locate(): base is /opt/Cantaloupe-3.3, name is cantaloupe.properties
14:38:21.592 [pool-2-thread-1] DEBUG o.a.c.c.DefaultFileSystem - Could not locate file cantaloupe.properties at /opt/Cantaloupe-3.3: no protocol: /opt/Cantaloupe-3.3
14:38:21.593 [pool-2-thread-1] DEBUG o.a.c.c.ConfigurationUtils - Loading configuration from the path /opt/Cantaloupe-3.3/cantaloupe.properties
14:38:21.593 [pool-3-thread-1] ERROR e.i.l.c.u.FilesystemWatcher$Callback - run(): null
14:38:21.593 [pool-3-thread-1] ERROR e.i.l.c.u.FilesystemWatcher$Callback - run(): null
14:16:34,058 |-INFO in ch.qos.logback.classic.LoggerContext[default] - Could NOT find resource [logback.groovy]
14:16:34,058 |-INFO in ch.qos.logback.classic.LoggerContext[default] - Could NOT find resource [logback-test.xml]
14:16:34,058 |-INFO in ch.qos.logback.classic.LoggerContext[default] - Found resource [logback.xml] at [file:/opt/apache-tomcat-8.0.36/webapps/ROOT/WEB-INF/classes/logback.xml]

4. I'm seeing errors regarding the log appender.

14:18:40,301 |-ERROR in ch.qos.logback.core.rolling.RollingFileAppender[AppRollingFileLog] - For more information, please visit http://logback.qos.ch/codes.html#earlier_fa_collision
...
14:24:06,403 |-ERROR in ch.qos.logback.core.rolling.RollingFileAppender[AppRollingFileLog] - 'File' option has the same value "/var/cantaloupe/cantaloupe.log" as that given for appender [AppFileLog] defined earlier.
14:24:06,403 |-ERROR in ch.qos.logback.core.rolling.RollingFileAppender[AppRollingFileLog] - Collisions detected with FileAppender/RollingAppender instances defined earlier. Aborting.
14:24:06,403 |-ERROR in ch.qos.logback.core.rolling.RollingFileAppender[AppRollingFileLog] - For more information, please visit http://logback.qos.ch/codes.html#earlier_fa_collision

5. Lastly, the properties file is reloaded twice again after.

I could send you the entire log if you want to. 

Thanks!

**Comment by adolski** [comment]
Hi @FrancoisPaquette,

Do you have multiple Cantaloupe instances using the same config file? In that case, they would both be writing to the same log file, and it would explain why the config file appears to be reloaded twice when changed.

Are your log.application.FileAppender.and RollingFileAppender both enabled and pointed to the same log file? I think it's OK to enable both of them, but they have to use different files, otherwise they will conflict. I think most people would probably only want to enable one or the other, though.

As for the `FilesystemWatcher$Callback` messages, I am able to get these to happen when I point multiple Cantaloupe instances to the same config file which has `delegate_script.enabled = true` and `delegate_script.pathname = ` (empty).

**Comment by FrancoisPaquette** [comment]
Hi,

There is only one instance running.  I tried the previous version again and now I'm seeing the config file also being reloaded, which wasn't the case before.  But there is no error for the FilesystemWatcher, though.  I'll investigate further, there is something weird going on.

Thanks

**Comment by FrancoisPaquette** [comment]
The development server looks fine. We have the same issue with the production one, I guess I didn't see it before.  I've ask one my coworker to take a look at it, maybe he'll see something.

As for the  FilesystemWatcher$Callback error and the settings you are refering to, this is what I have in my config file:

delegate_script.enabled = true
delegate_script.pathname = delegates.rb

I'm getting this error with the version 3.3. and not 3.2.2.

**Comment by adolski** [comment]
It looks like there are a few different things going on here. I will just add my thoughts on each individual thing that I can see.

1. The timestamps "jumping into the past" like that makes me think that there must be two application instances, running on different machines, with different system clocks, writing to the same log file. I don't know how it's possible for the timestamps to vary so much otherwise.
2. The properties file being reloaded twice after being changed -- are you sure it's being reloaded twice, or are there two instances, each reloading it once, but both logging it to the same file, so it looks like twice? I know you've already said there is only one instance, but this is the only thing I can think of. 
3. The `FilesystemWatcher$Callback` messages are actually coming from the delegate script watcher, not the config watcher.
    1. It makes sense that they don't appear in 3.2 because the delegate script watcher was not present in that version.
    2. I admit that these are not very helpful messages, and I will clean that up in a future release.
    3. They may be caused by a FileNotFoundException involving the delegate script. Since you have set `delegate_script.pathname` to `delegates.rb`, does delegates.rb exist, in the same folder as the config file?
4. The messages from `ch.qos.logback.classic.LoggerContext[default]` are generated internally by the logger and are harmless.
5. The messages about the collisions are explained in my previous comment -- FileAppender and RollingFileAppender can't write to the same file simultaneously.

**Comment by FrancoisPaquette** [comment]
Sorry for the delay. I'm able to recreate the issue with the differents timestamps, it happens when when the FileAppender and RollingFileAppender are writing to the same file. 

I'm not getting errors anymore either but it still look like the properties file is loaded twice.  It doesn't happen when the tomcat is restarted.

Reloading config file: /opt/cantaloupe/cantaloupe.properties
08:43:54.427 [pool-2-thread-1] DEBUG o.a.c.c.PropertiesConfiguration - FileName set to cantaloupe.properties
08:43:54.431 [pool-2-thread-1] DEBUG o.a.c.c.PropertiesConfiguration - Base path set to /opt/cantaloupe
08:43:54.432 [pool-2-thread-1] DEBUG o.a.c.c.ConfigurationUtils - ConfigurationUtils.locate(): base is /opt/cantaloupe, name is cantaloupe.properties
08:43:54.433 [pool-2-thread-1] DEBUG o.a.c.c.DefaultFileSystem - Could not locate file cantaloupe.properties at /opt/cantaloupe: no protocol: /opt/cantaloupe
08:43:54.434 [pool-2-thread-1] DEBUG o.a.c.c.ConfigurationUtils - Loading configuration from the path /opt/cantaloupe/cantaloupe.properties
Reloading config file: /opt/cantaloupe/cantaloupe.properties
08:43:54.521 [pool-2-thread-1] DEBUG o.a.c.c.PropertiesConfiguration - FileName set to cantaloupe.properties
08:43:54.522 [pool-2-thread-1] DEBUG o.a.c.c.PropertiesConfiguration - Base path set to /opt/cantaloupe
08:43:54.522 [pool-2-thread-1] DEBUG o.a.c.c.ConfigurationUtils - ConfigurationUtils.locate(): base is /opt/cantaloupe, name is cantaloupe.properties
08:43:54.523 [pool-2-thread-1] DEBUG o.a.c.c.DefaultFileSystem - Could not locate file cantaloupe.properties at /opt/cantaloupe: no protocol: /opt/cantaloupe
08:43:54.523 [pool-2-thread-1] DEBUG o.a.c.c.ConfigurationUtils - Loading configuration from the path /opt/cantaloupe/cantaloupe.properties

It looks like the file is being located correctly also.  I'm absolutely sure there is only one instance running. 

Thanks

**Comment by adolski** [comment]
It looks like the config file being reloaded twice is the only remaining issue here. I still can't get it to happen on my end, though.

The reloading feature uses the Java [WatchService](https://docs.oracle.com/javase/8/docs/api/index.html?java/nio/file/WatchService.html). The docs say:

> The implementation that observes events from the file system is intended to map directly on to the native file event notification facility where available, or to use a primitive mechanism, such as polling, when a native facility is not available. Consequently, many of the details on how events are detected, their timeliness, and whether their ordering is preserved are highly implementation specific. For example, when a file in a watched directory is modified then it may result in a single ENTRY_MODIFY event in some implementations but several events in other implementations.

So my next guess is that last sentence is what's happening in your environment. For me (macOS) there is only one file-modified event sent, but perhaps your environment sends two events?

I think modifying the reloading behavior to skip reloading if the last reload was less than X seconds ago would fix this.

**Comment by adolski** [comment]
If there are indeed multiple file-modified events being sent, the above commit should cause the configuration to only be reloaded once.

**Comment by FrancoisPaquette** [comment]
I'm doing a bit of reasearch regarding this class and and it looks like this is a well-known bug.  Someone on another forum is suggesting that the event is triggered when the file is opened and again when it is saved. It could be it.

**Comment by adolski** [comment]
Very possible. If opening the file updates its last-accessed time, then maybe that could trigger a file-modified event?

**Comment by devanonyme** [comment]
I checked this one. I'm able to reproduce the double reload on windows with Cantaloupe Standalone 3.3

I was thinking at first that if there was a file access timestamp that was raising a change event and not enought granularity to filter it that we could check the content hash. That would be inefficient for a full folder but probably good enough for a config file.

It would be better to find a solution so that the file stream is only read once but I hacked a quick test inside PropertiesConfiguration.java

1) Added a private : private String md5_previous_file_read = "";

2) Added this function in public synchronized void reload() {

            try {
                FileInputStream fileContentForHashCalculation = new FileInputStream(configFile);
                String md5 = org.apache.commons.codec.digest.DigestUtils.md5Hex(fileContentForHashCalculation);
                fileContentForHashCalculation.close();
                
                if (md5.equals(md5_previous_file_read)) {
                    return;
                } else {
                    md5_previous_file_read = md5;
                }
            } catch (IOException e) {
                // ... 
            }  

And then I just change "log.application.level = warn" in config file alternating to debug and  warn by exemple and observe the result.

The hack does it job by skipping the second reload but I also notice times where the reload is done twice and even other times where the double reload doesnt occur.

There is maybe something else happening, I'll report if I find something else.

**Comment by devanonyme** [comment]
Just an exemple of the weird behavior, it's something like
log.application.level = all
I change to debug.. 
I get two reload one skip
I change to all
I get two reload no skip

Maybe there are other folder watch that could cause conflict? (my logs folder is in cantaloupe folder so technically at the same level as config file... I'll try to move the logs out and see.

**Comment by adolski** [comment]
I was able to reproduce the double-reloading in Windows 7. I like your idea of checking the hash value of the file contents, so I implemented that and it seems to solve it for me, though my only testing was to change the file a few times in Windows. Strange that you are getting different behavior after making the same change...

**Comment by devanonyme** [comment]
Cool. I plan to investigate a little more just to understand what could cause the varying behavior (even with hash validation), I let you know if I find something later.
 
It's not high priority for us (as we are not suppose to switch flags at will in production) but did notice today that a reload didn't work as expected (with the delegate file this time). There was an error in the script that we noticed while testing our release of 3.3 in production, I corrected it and it looks like it ran the older script when fullfulling the request (the first one, the second attempt was ok).
